tap_recharge/streams.py

```diff
@@ -0,0 +1,443 @@
+"""
+This module defines the stream classes and their individual sync logic.
+"""
+
+import datetime
+
+from typing import Iterator
+
+import singer
+from singer import Transformer, utils, metrics
+
+from tap_recharge.client import RechargeClient
+
+
+LOGGER = singer.get_logger()
+
+MAX_PAGE_LIMIT = 250
+
+
+class BaseStream:
+    """
+    A base class representing singer streams.
+
+    :param client: The API client used to extract records from external source
+    """
+    tap_stream_id = None
+    replication_method = None
+    replication_key = None
+    key_properties = []
+    valid_replication_keys = []
+    path = None
+    params = {}
+    parent = None
+    data_key = None
+
+    def __init__(self, client: RechargeClient):
+        self.client = client
+
+    def get_records(self,
+                    bookmark_datetime: datetime = None,
+                    is_parent: bool = False) -> list:
+        """
+        Returns a list of records for that stream.
+
+        :param bookmark_datetime: The datetime object representing the
+            bookmark date
+        :param is_parent: If true, may change the type of data
+            that is returned for a child stream to consume
+        :return: list of records
+        """
+        raise NotImplementedError("Child classes of BaseStream require "
+                                  "`get_records` implementation")
+
+    def get_parent_data(self, bookmark_datetime: datetime = None) -> list:
+        """
+        Returns a list of records from the parent stream.
+
+        :param bookmark_datetime: The datetime object representing the
+            bookmark date
+        :return: A list of records
+        """
+        # pylint: disable=not-callable
+        parent = self.parent(self.client)
+        return parent.get_records(bookmark_datetime, is_parent=True)
+
+
+# pylint: disable=abstract-method
+class IncrementalStream(BaseStream):
+    """
+    A child class of a base stream used to represent streams that use the
+    INCREMENTAL replication method.
+
+    :param client: The API client used extract records from the external source
+    """
+    replication_method = 'INCREMENTAL'
+
+    # pylint: disable=too-many-arguments
+    def sync(self,
+            state: dict,
```

Comment:
There's a lot of these alignment issues in this file. Does your editor allow you to indent everything at once?

Reply:
I was trying to match the indent style of the rest of the original tap code, but have adjusted the function/method definitions and calls to match PEP8 indentation.

Reply:
I see the way this was resolved and I'm okay with it.

I originally saw a lot of 
```python
def sync(self,
        state):
    pass

def sync(self,
          state):
    pass
```

Both of which I thought you would resolve to
```python
def sync(self,
         state):
    pass
```

The real solution here is to run a code formatter, but that's too far down my todo list